python-package/lightgbm/callback.py

```diff
@@ -108,24 +111,30 @@ def reset_learning_rate(learning_rates):
         current number of round and the total number of boosting round \
         (e.g. yields learning rate decay)
         - list l: learning_rate = l[current_round]
-        - function f: learning_rate = f(current_round, total_boost_round)
-
+        - function f: learning_rate = f(current_round, total_boost_round) \
+                   or learning_rate = f(current_round)
     Returns
     -------
     callback : function
         The requested callback function.
     """
     def callback(env):
         """internal function"""
-        booster = env.model
-        iteration = env.iteration
         if isinstance(learning_rates, list):
-            if len(learning_rates) != env.end_iteration:
-                raise ValueError("Length of list 'learning_rates' has to equal 'num_boost_round'.")
-            booster.reset_parameter({'learning_rate':learning_rates[iteration]})
+            if len(learning_rates) != env.end_iteration - env.begin_iteration:
+                raise ValueError("Length of list 'learning_rates' has to equal to 'num_boost_round'.")
+            env.model.reset_parameter({'learning_rate':learning_rates[env.iteration]})
         else:
-            booster.reset_parameter({'learning_rate':learning_rates(iteration, env.end_iteration)})
+            argc = len(inspect.getargspec(learning_rates).args)
```

Comment:
ok, I think we can according to this to support better fobj and feval. We can do this in another PR.